File: android/src/main/java/com/scalesforgrams/plugins/arscanner/ARScannerPlugin.java
package com.scalesforgrams.plugins.arscanner;

import android.Manifest;
import android.os.Handler;
import android.os.Looper;
import android.webkit.WebView;
import androidx.lifecycle.LifecycleOwner;
import com.getcapacitor.JSObject;
import com.getcapacitor.PermissionState;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;

@CapacitorPlugin(name = "ARScanner", permissions = { @Permission(strings = { Manifest.permission.CAMERA }, alias = "camera") })
public class ARScannerPlugin extends Plugin {

    private static final String CAMERA_ALIAS = "camera";

    private CameraPreviewManager previewManager;
    private final Handler mainHandler = new Handler(Looper.getMainLooper());

    @PluginMethod
    public void checkSupport(PluginCall call) {
        JSObject result = new JSObject();
        result.put("isSupported", true);
        result.put("hasLidar", false);
        result.put("hasDepthApi", false);
        result.put("depthQuality", "estimated");
        call.resolve(result);
    }

    @PluginMethod
    public void startPreview(PluginCall call) {
        if (getPermissionState(CAMERA_ALIAS) != PermissionState.GRANTED) {
            requestPermissionForAlias(CAMERA_ALIAS, call, "cameraPermissionCallback");
            return;
        }
        doStartPreview(call);
    }

    @PermissionCallback
    private void cameraPermissionCallback(PluginCall call) {
        if (getPermissionState(CAMERA_ALIAS) == PermissionState.GRANTED) {
            doStartPreview(call);
        } else {
            call.reject("Camera permission denied");
        }
    }

    private void doStartPreview(PluginCall call) {
        mainHandler.post(() -> {
            try {
                WebView webView = getBridge().getWebView();
                LifecycleOwner lifecycleOwner = (LifecycleOwner) getActivity();

                if (previewManager != null) {
                    previewManager.stop(webView);
                }

                previewManager = new CameraPreviewManager();
                // Resolve only when CameraX is actually bound: resolving early made the JS
                // side mark the preview as running 1-3s before it really was, and every
                // capture() in that window failed with "Preview is not running".
                previewManager.start(webView, lifecycleOwner, new CameraPreviewManager.StartCallback() {
                    @Override
                    public void onStarted() {
                        JSObject result = new JSObject();
                        result.put("started", true);
                        call.resolve(result);
                    }

                    @Override
                    public void onError(String message) {
                        notifyScanError("Failed to start camera: " + message);
                        call.reject("Failed to start camera: " + message);
                    }
                });
            } catch (Exception e) {
                call.reject("Failed to start preview: " + e.getMessage());
            }
        });
    }

    /**
     * Mirror of the iOS scanEvent error contract so the web layer's capture issue
     * tracking works on Android too (previously Android emitted no scanEvents at all).
     */
    private void notifyScanError(String message) {
        JSObject event = new JSObject();
        event.put("type", "error");
        event.put("message", message);
        notifyListeners("scanEvent", event);
    }

    @PluginMethod
    public void stopPreview(PluginCall call) {
        mainHandler.post(() -> {
            if (previewManager != null) {
                WebView webView = getBridge().getWebView();
                previewManager.stop(webView);
                previewManager = null;
            }
            JSObject result = new JSObject();
            result.put("stopped", true);
            call.resolve(result);
        });
    }

    @PluginMethod
    public void capture(PluginCall call) {
        if (previewManager == null || !previewManager.isRunning()) {
            call.reject("Preview is not running. Call startPreview() first.");
            return;
        }

        previewManager.capture(
            new CameraPreviewManager.CaptureCallback() {
                @Override
                public void onResult(JSObject result) {
                    call.resolve(result);
                }

                @Override
                public void onError(String message) {
                    notifyScanError(message);
                    call.reject(message);
                }
            }
        );
    }

    @PluginMethod
    public void setTorch(PluginCall call) {
        boolean enabled = call.getBoolean("enabled", false);

        if (previewManager == null || !previewManager.isRunning()) {
            call.reject("Preview is not running");
            return;
        }

        mainHandler.post(() -> {
            boolean result = previewManager.setTorch(enabled);
            JSObject res = new JSObject();
            res.put("enabled", result);
            call.resolve(res);
        });
    }
}

File: ios/Sources/ARScannerPlugin/ARScannerPlugin.swift
import AVFoundation
import Capacitor
import UIKit

@objc(ARScannerPlugin)
public class ARScannerPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "ARScannerPlugin"
    public let jsName = "ARScanner"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "checkSupport", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "startPreview", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "stopPreview", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "capture", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setTorch", returnType: CAPPluginReturnPromise),
    ]

    private let implementation = ARScanner()
    private var previewManager: CameraPreviewManager?

    // MARK: - checkSupport

    @objc public func checkSupport(_ call: CAPPluginCall) {
        let hasLidar = implementation.isLidarAvailable()
        let isSupported = implementation.isARSupported()

        let depthQuality: String
        if hasLidar {
            depthQuality = "high"
        } else if isSupported {
            depthQuality = "low"
        } else {
            depthQuality = "none"
        }

        call.resolve([
            "isSupported": isSupported,
            "hasLidar": hasLidar,
            "hasDepthApi": false,
            "depthQuality": depthQuality
        ])
    }

    // MARK: - startPreview (camera behind WebView)

    @objc public func startPreview(_ call: CAPPluginCall) {
        // Gate on camera permission BEFORE running the AR session
        switch AVCaptureDevice.authorizationStatus(for: .video) {
        case .authorized:
            doStartPreview(call)
        case .notDetermined:
            AVCaptureDevice.requestAccess(for: .video) { [weak self] granted in
                if granted {
                    self?.doStartPreview(call)
                } else {
                    call.reject("Camera permission denied")
                }
            }
        default: // .denied, .restricted
            call.reject("Camera permission denied")
        }
    }

    private func doStartPreview(_ call: CAPPluginCall) {
        let forceLidarOff = call.getBool("forceLidarOff") ?? false

        DispatchQueue.main.async { [weak self] in
            guard let self = self,
                  let webView = self.webView else {
                call.reject("WebView not available")
                return
            }

            // Stop existing preview if any
            self.previewManager?.stop()

            let manager = CameraPreviewManager()
            manager.forceLidarOff = forceLidarOff
            manager.onEvent = { [weak self] event in
                self?.notifyListeners("scanEvent", data: event)
            }

            manager.start(webView: webView)
            self.previewManager = manager

            call.resolve(["started": true])
        }
    }

    // MARK: - stopPreview

    @objc public func stopPreview(_ call: CAPPluginCall) {
        DispatchQueue.main.async { [weak self] in
            self?.previewManager?.stop()
            self?.previewManager = nil
            call.resolve(["stopped": true])
        }
    }

    // MARK: - capture (measure from live preview)

    @objc public func capture(_ call: CAPPluginCall) {
        guard let manager = previewManager, manager.isRunning else {
            call.reject("Preview is not running. Call startPreview() first.")
            return
        }

        manager.capture { result in
            if let result = result {
                call.resolve(result)
            } else {
                call.reject("Capture failed — try again")
            }
        }
    }

    // MARK: - setTorch

    @objc public func setTorch(_ call: CAPPluginCall) {
        let enabled = call.getBool("enabled") ?? false

        DispatchQueue.main.async { [weak self] in
            guard let manager = self?.previewManager, manager.isRunning else {
                call.reject("Preview is not running")
                return
            }
            let result = manager.setTorch(enabled: enabled)
            call.resolve(["enabled": result])
        }
    }
}
